### PR TITLE
Fixing EOL replace. Fixes #54

### LINKED ISSUE #54: Enabling editorconfig plugin scrambles my files
After upgrade to atom 1.1, I got all my files in editor somehow scrambled and realigned.
When analyzing the issue, I have figured out that this happens only when I enable the `editorconfig` package (if the package is disabled, it doesn't happen).

Here is a screenshot of my file with `editorconfig` disabled:
atom-pre

And again with `editorconfig` package enabled:
atom-post

My `.editorconfig` file is:

		# EditorConfig is awesome: http://EditorConfig.org

		# top-most EditorConfig file
		root = true

		# Unix-style newlines with a newline ending every file
		[*]
		end_of_line = lf
		insert_final_newline = true
		indent_style = tab
		indent_size = 4
		charset = utf-8

		# 4 space indentation
		[*.{py}]
		indent_style = space
		indent_size = 4

		# 2 space indentation
		[*.{js,jsx}]
		indent_size = 2
		indent_style= space

		# Indentation override for all JS under lib directory
		#[lib/**.js]
		#indent_style = space
		#indent_size = 2

		# Matches the exact files either package.json or .travis.yml
		#[{package.json,.travis.yml}]
		#indent_style = space
		#indent_size = 2

### PR DESCRIPTION
**Pull request by vmuriart**
Tested out the original issue from #51 and this fix doesn't re-open the it.

-

Fixes #54.

### CODE REVIEW

**Comment by sindresorhus** [comment]
// @pkese @dwelle

**Comment by dwelle** [comment]
Doesn't fix. Here's me redo-ing what the replace does:

crlf_bug

It seems the `/\r?\n/g` replaces stuff that it shouldn't, so there's certainly bug in `scanInRange` or its replace function.

**Comment by vmuriart** [comment]
@dwelle, just to make sure i understood, the patch above didn't fix. 
But reverting back to this does correct?
```js
        if (config.end_of_line && config.end_of_line in lineEndings) {
            const preferredLineEnding = lineEndings[config.end_of_line];
            const buffer = editor.getBuffer();
            buffer.setPreferredLineEnding(preferredLineEnding);
            buffer.setText(buffer.getText().replace(/\r?\n/g, preferredLineEnding));
        }
```

**Comment by dwelle** [comment]
@vmuriart correct

**Comment by vmuriart** [comment]
In your case, from which line ending are you going, and to which line ending are you going? The case i tested is `/n` to `/r/n` which seemed to be the general case that @pkese reported to be the issue (when line ending lengths are different).

**Comment by vmuriart** [comment]
fair warning, my experience with lineendings is very limited, so forgive me if it sounds somewhat ignorant on my side :smile:

**Comment by vmuriart** [comment]
tl;dr: [This](https://github.com/atom/text-buffer/blob/master/src/text-buffer.coffee#L672) may be used as an example to write the correct patch

This patch doesn't work when going from `CRLF` to `LF`. 
At this point this `pr` isn't the correct fix to address both issues #54 without reverting #51. 

Below going to document my findings so far in case it gives anyone else a direction to work on.

The underlying issues with `buffer.setText(buffer.getText().replace(/\r?\n/g, preferredLineEnding));` is that [`setText`](https://github.com/atom/text-buffer/blob/master/src/text-buffer.coffee#L560) destroys the and recreates the buffer with the next text and in the process loses the position of the cursor. 
I tried extracting the position of the cursor before `setText` is executed by using [go-to-line](https://github.com/atom/go-to-line/blob/master/lib/go-to-line-view.coffee) as a reference, but always returns to the top row instead of the called line 

```js
		if (config.end_of_line && config.end_of_line in lineEndings) {
			const preferredLineEnding = lineEndings[config.end_of_line];
			const buffer = editor.getBuffer();
			const currentRow = editor.getCursorBufferPosition().row;
			buffer.setPreferredLineEnding(preferredLineEnding);
			buffer.setText(buffer.getText().replace(/\r?\n/g, preferredLineEnding));
			editor.setCursorBufferPosition(currentRow);
		}
``` 


The problem with ` buffer.backwardsScanInRange(/\r?\n/g, ...);` is that we are pretty much just shooting in the dark hoping that it works without really understanding what its meant to do. Contrary to my `pr` it seems that [`backwardsScanInRange`](https://github.com/atom/text-buffer/blob/master/src/text-buffer.coffee#L1016) was the correct function to use but it isn't doing what we *think* it should do. 

The one that I thought would do it for us was going to be `buffer.replace(/\r?\n/g, preferredLineEnding));` but even on a function that should be pretty standard on its application, the output wasn't what I had expected and still returned *funky* text. This may indicate that there is something wrong with the `buffer` and how it does `replace`. 

While writing all this, I noticed the [applyChange](https://github.com/atom/text-buffer/blob/master/src/text-buffer.coffee#L672) that seems to be able to normalize the lineendings as we want to.

As a side note, I think the issue wasn't introduced by Atom 1.1 release, but rather by the fix that was introduced for #51. I'll try working on this again later next week. @sindresorhus feel free to close this `pr` if you feel it appropriate to do so.

**Comment by dwelle** [comment]
My case was from `LF` (`\n`) to `CRLF` (`\r\n`). I'd be happy to look into this, but am under time pressure, too, so not sure when I get the chance.

**Comment by sindresorhus** [comment]
Closing in favor of #61.
